### ISSUE TITLE
Freqtradebot.py refactoring and Orders 

### ISSUE THREAD

**Issue by mishaker** [freqtrade/freqtrade#1335]
This is just to explain my thoughts on refactoring freqtradebot.py file, it is becoming too fat. 
The way I see different categories of its functionalities:

## Orders 
Think of it as an API, the return will be success or fail. it contains everything related to an order: order type, order time in force, etc. it takes an input, it then sends to an exchange, end of.

## Order execution management 
This is the component responsible for "what happens if?". what happens if the limit order fulfilled partially ? should the bot cancel the other part ? should the bot change the price so it would get more chance to be fulfilled ?
Or maybe you want to put the stoploss immediately after the buy order is fulfilled. but maybe not ...
Ideally this part should be controlled by the strategy.But I suspect that would be a mess. So probably another component ...

## Exchange
Everything related to exchange interaction. there should be no exchange related information in freqtradebot.py. The interaction with any exchange should be customizable. freqtrade should just call exchange.py. exchange in turn, should be an encapsulation and overridable CCXT library.

## Freqtrade 
It does everything related to coordination. purely technical. nothing related to trading. it is in charge of interaction with other components, catching errors, DB connections, etc.

**Comment by xmatthias** [comment]
i tend to agree (at least partially) ... `freqtradebot.py` is bloated, WAY too large, and needs to be split up somehow ...

You need to be carefull however, as your main focus seems to be to make "everything" overridable, which "everything" in python is anyway, i could subclass `Exchange` right now, override one method and add that in 2 lines (1 import, one where it's actually instanciated) to the bot. This should not be a focus at all, as it is the very last step - and VERY easy to accomplish quite dynamically (see how strategyresolver / dynamic hyperopt in #571) - even without abstract classes.

### OrderHandler
what you call "order execution management" - i would call "OrderHandler".
The bot passes it an order (type, amount, ...) , and it takes care of "whatever" necessary, including timeout handling, cancelling etc.

This combines (in theory) the 2 steps you have above about orders and order executions - but details if 2 seperate classes make sense are not clear to me yet (i don't currently see the benefit, but didn't look at it in detail yet).

### Exchange
we got that already - and nothing is using ccxt directly without this class - so not sure what you mean by that ... 

### Dynamic whitelist

Just an idea so far - but i think extracting dynamic whitelist based on different factors (volume, postitive volume, ...) would be benefitial. Our current algorithm is not that good (it'll only look at volume - so will jump on the coin loosing the most too), however i wouldn't just change it, but offer alternatives (`Edge` beeing one, but there are multiple other ways to determine which coins to anlayze, which may or may not work depending on the strategy used ...)

### freqtradebot
coordinates everything, gets buy/sell signals from strategy, passes that to `OrderHandler` for "order handling" ...

**Comment by mishaker** [comment]
**General comment:** I think there is a misunderstanding around "abstract". Let's forget completely about abstract classes. After seeing you pointed to #571, I think maybe you didn't read my proposal on #1326. Basically what I suggested came to "resolver" approach. Maybe I went too far by putting an abstract class on top of it but the idea was having a Default class which can be overridden by the resolver (same as in #571).

### OrderHandler
I am ok with the naming but not a fan of your way of encapsulation. Take "order cancelling" as an example. You would put it in  `freqtradebot.py` but order cancelling depends also on order types: If my order type is stoploss on exchange then freqtradebot should cancel the stoploss before selling. You could argue that it is easier to leave it there but I would suggest to delegate everything to OrderHandler. freqtradebot should be responsible only for calling other components, logs, etc (purely technical, it should be neutral trading wise.
IMO nothing should prevent us to have a default OrderHandler which can be overridden <br/> by resolver(i.e. --OrderHandler=MyNastyOrders)

### DataProvider
"Context"  is more understandable. It is used in zipline too and people are familiar with it and I don't see any reason to innovate here (see https://github.com/freqtrade/freqtrade/issues/1192#issuecomment-419683661). It should be global and accessible by all components.

### Exchange
We have that already but it is becoming dirty with all those if 'biannce' then up else down ...
In theory I don't see any reason to patch it via a resolver and specifying it before running the bot (same mechanism as described in OrderHandler) 

### Dynamic whitelist
As I said multiple times before, dynamic whitelist is at best misleading if not useless. I searched about how you can distinguish between sell volume and buy volume: there is no way. Volume is volume and there is no way to break it down to sell and buys. That being said, I would like to ask help from @creslinux on this point as I am not a 100% sure. if Creslin confirms it, then I would personally vote for ripping it off honestly.

### Freqtradebot 
That's the part responsible of all you mentioned. It is already a huge task: Thread managing, async, error handling, etc, etc. I would not put a bit of trading logic in this file.

**Comment by xmatthias** [comment]
### Dynamic whitelist
you can't distinguish it, no, but you do have data to determine if it went up or down in the last X days - if it went up in the last 7 days, but the logic behind what type of algorithm should be used here shall not be part of this issue.

### OrderHandler
> You would put it in freqtradebot.py but order cancelling depends also on order types:

not sure where you got that from - but not from me. 
my comment above clearly states (*no edit since yesterday evening - so i didn't change anything*) that `OrderHandler` does "whatever necessary, including cancellation": 

> The bot passes it an order (type, amount, ...) , and it takes care of "whatever" necessary, including timeout handling, cancelling etc.

### Exchange
I'm not sure about that - we may have to in the long run, but i currently don't see a need to subclass exchange for exchange X. the "`if 'binance'`" is only there once currently - unless you plan to add that to every other function, then refactoring that to a seperate class is overkill. Don't make it too complicated until we really have to - 
another class to maintain ... just because 1 or 2 if statements? no thank you.

### DataProvider
Names are not important, the functioning is, we can even call it "GetMeProfit" if you prefer that name... 

#1326 tried to do 2 things at once (allow market orders, refactor orders to a class) - which makes it confusing at best.

**Comment by creslinux** [comment]
Hi, 
 question had me tagged re volume. 
FT is only storing part of the kline(candle) data OHLCV, with V Volume being a total 

The candle also includes a break down in to volume agains the ASK book (buying longs, the base currency) or taker against the BID book, selling into bids, quote currency. 

This tells you if volume is from buying up or selling down 

```
[
  [
    1499040000000,      // Open time
    "0.01634790",       // Open
    "0.80000000",       // High
    "0.01575800",       // Low
    "0.01577100",       // Close
    "148976.11427815",  // Volume
    1499644799999,      // Close time
    "2434.19055334",    // Quote asset volume
    308,                // Number of trades
    "1756.87402397",    // Taker buy base asset volume
    "28.46694368",      // Taker buy quote asset volume
    "17928899.62484339" // Ignore.
  ]
]
```

**Comment by creslinux** [comment]
.. thats binance specific.  Which seems the purpose of this thread.

**Comment by xmatthias** [comment]
### Exchange
subclassing exchange will take a few days of work, where adding a "if" does not.

Extracting conditional statements per exchange into a subclass will be a much simpler task once the needed logic is in place with if-statements.

You seem to mention incompatibility between exchanges and the need to subclass - but for now, we have NO feature which would require this, so it's bloating the bot up with something currently not needed. 
If you read carefully, then i'm not opposing subclassing exchange in the longer term, but i think that time is much better spent short-term in adding the functionality you need (i assume it's stoploss based on #1341) than to architect it perfectly and extensibly, which we can do in the future - but for now it would add 0 benefit, as we have nothing you could overwrite anyway. 

#1341 can be accomplished with conditionals (i assume in the first step, it's going to be binance only?). After that, we can think if it makes sense at that point.

### DataProvider
Don't feel upset if i don't really mind about the naming ... Context seems to come from zipline - looking at Gekko or zenbot, there is no context as far as i could find - and in contrast to zipline gekko/zenbot do support live trading - so i wouldn't call Context "established culture" just because of 1 project ... :smile:

**Comment by mishaker** [comment]
### Exchange
Ok, let's move on. for now I am going to implement #1341 in exchange.py then we will see later.

### DataProvider 
Ok :) then please draft something. on #1338 or somewhere else.

I would keep this issue open for future ref.

**Comment by creslinux** [comment]
For what it's worth, abstracting and sub-classing seems cleaner to me, it
gets the pattern right

A couple else ifs now will become many soon, and are inherently inline
where could impact another exchange. They're psuedo global given the flow
of FT

Add another exch for even more else ifs it will be unmaintainable.

Easier to call the subclass that's relevant than waterfall if else branches
for all calls all times

**Comment by xmatthias** [comment]
@creslinux 
in theory and long-term i agree, and this is what it'll probably/eventually move to. however, i think it's a step-by-step process, and it's more important (I think, anyway) to have the functionality working correctly, and THEN refactoring already working stuff to have a globally applicable solution.
We can also spend a few weeks now to subclass exchange in a proper way ... and restructure the architecture, but i would prefer to have the functionality in in a "working" way sooner ...

extracting 3 `if`'s when necessary is easier and a lot quicker than to abstract it now, before we even need what exactly the needs are - which we'll (hopefully) see in/with #1341

**Comment by mishaker** [comment]
There is another problem that we have in order execution sequence which is also addressed in #1362:

Let's say the interval is 5 min, and bot receives a buy signal. it then tries to buy at the ticker price (or orderbook). for some reason the buy is not fulfilled. it then tries the same after an iteration but with new ticker, this process continues till: either on next interval buy signal disappears or the buy is fulfilled. 
This is dangerous, as the price can move against you and currently the bot doesn't consider the slippage. 
There should be a mechanism available to the strategy (or somewhere else) to handle it.
As an example we can imagine the following: 

- A signal is indexed and uniquely identified by the last close candle and pair.
- Number of tries and reasons of success or failure is saved somewhere for that particular signal.
- Decisions are made based on previous failure/success reasons by the strategy. 

I am unsure of where we can implement it in current architecture (specially because the current architecture is prone to change) 

But the first solution which comes to mind is to relate order client id to signal id. it is then passed to the strategy: 

If signal.id in orders and orders[signal.id] == "REJECTED" then ... 

It is not mature enough what I am saying, but I would like to open the discussion about it.

**Comment by xmatthias** [comment]
you can disable this by running TA "on candle" - so the buy-signal disappears after the initial signal. 

Check the documentation how to configure that ... it's in strategy if i remember correctly.

If you think that's not enough please open a new issue - otherwise it'll get messy as hell in this issue, which combines 5 different topics already...

**Comment by iuvbio** [comment]
Hi all,

apparently I didn't do my research properly before opening #1563, sorry about that. For what it's worth, here my 2 cents:

Exchange
---
> Basically what I suggested came to "resolver" approach. Maybe I went too far by putting an abstract class on top of it but the idea was having a Default class which can be overridden by the resolver (same as in #571).

This is exactly what I did. I think it's a good idea to have a base class (not ABC) that works with any exchange "out of the box", and a subclass for each new exchange that is considered supported for more advanced use. This would also greatly help development, as it's easier for people who might have knowledge about a particular exchange to work on that without having to change the base class. Besides, the huge number of exchanges supported by `ccxt` makes it infeasible to have a subclass for each one and like this FT can still theoretically be used with "unsupported" exchanges (as it is now).
I also agree with @xmatthias that having the functionality in a "working" way before adding additional stuff. What exactly is still "not working" at the moment? Looking through the open issues, most seem to be features or minor bugs. I think the base functionality is working rather well.

Dynamic Whitelist
---
In my opinion, this is a good feature and should stay, even when the exchange doesn't allow to distinguish between buy and sell volume. It provides a certain "protection" against the bot trading pairs that have very low volume and therefore an order is more likely not to be filled. What really should be addressed is the pairlist selecting coins where the price is only 2 digits above precision (not sure if I term that correctly). With `stoploss_on_exchange` enabled, this currently leads to the bot breaking down completely, which is the worst possible situation (#1520). I'm not in favour of the solution suggested there (increasing the stoploss by an arbitrary amount for that trade), but would rather remove the pair from the validated whitelist or even add it to the blacklist. Just not sure whether that should be part of the current dynamic pairlist, its own pairlist class, or just some logic that trigger when `stoploss_on_exchange` is even used.

**Comment by xmatthias** [comment]
the "working" was mainly pointed at stoploss_on_exchange, which at the time of writing was not completely implemented -> not working correctly yet.


Whitelists can have parameters [see here](https://www.freqtrade.io/en/latest/configuration/#dynamic-pairlists) ... adding one to filter out "low digit" stuff should be easy to implement into the current whitelist.

However if you have another idea on how to "sort/filter" a whitelist a new class is also welcome - but I think it should do something distinguishible from the current dynamic whitelist.

**Comment by iuvbio** [comment]
> Whitelists can have parameters [see here](https://www.freqtrade.io/en/latest/configuration/#dynamic-pairlists) ... adding one to filter out "low digit" stuff should be easy to implement into the current whitelist.

Yes it can be done, my question is rather about whether it should go there, as then the pairlist woudn't filter on volume only anymore, so the name might be misleading.

**Comment by iuvbio** [comment]
About exchanges..

I think one of the issues with this whole subclassing thing is that we'd essentially be doing ccxt's work. In theory, stoploss orders etc. should be unified there and I think they will be at some point, so all the work we'll have done on it would be rendered useless. If we could somehow collaborate with @kroitor though and he could make use of the work done here, that could be very fruitful. I think I saw in some issue that someone had spoken to him.. does anyone have contact with him?

For now though, if we're going the route of subclassing, I think there should be a boolean exchange attribute that indicates whether it can be used and the validation based on that, instead of specifically checking Binance. I guess the stoploss_limit method in it's current form should be moved to a Binance subclass as well, unless the same would work on other exchanges too.

**Comment by mishaker** [comment]
There are things which are not unifiable or at least very hard or as far as I know not a part of ccxt roadmap (ex. leverage orders). also ccxt doesn't support all exchanges. subclassing will allow user to implement a API on top of it. ccxt will remain in FT as it facilitates considerably API handling but it doesn't mean we should not give the possibility to the user to implement further functionality of an exchange.

**Comment by kroitor** [comment]
@mishaker 

> not a part of ccxt roadmap (ex. leverage orders)

Actually, it's on our roadmap, along with margin/positional trading and derivatives. We are developing our understanding of how to unify it properly. Not an easy thing to do, though...

> also ccxt doesn't support all exchanges.

We'll get cracking! )

@iuvbio 

> If we could somehow collaborate with @kroitor though and he could make use of the work done here, that could be very fruitful. I think I saw in some issue that someone had spoken to him.. does anyone have contact with him?

I've heard that's nearly impossible...

**Comment by xmatthias** [comment]
> I think one of the issues with this whole subclassing thing is that we'd essentially be doing ccxt's work

yes - but it's easier to do it for 3 selected exchanges, than for 100+ in 3 languages...
And once it's unified in ccxt, we can remove our "helper" code and have support for all exchanges ;) 

Yes, a binance subclass should be created as well, extracting all the "binance" specific logic.
Even with the best unification, some exchange-specifics will probably always remain, since not all exchanges support everything (just look at the "agreement" case for kraken).

> I think there should be a boolean exchange attribute _for time in force_

yes - i was thinking of implementing something similar to ccxt's `has` dict - (maybe `ft_has`) - which defines which subclass supports what feature (can be right next to the checks we do against the `has` dict...

> also ccxt doesn't support all exchanges. subclassing will allow user to implement a API on top of it

If a exchange is missing from ccxt, it's honestly better to submit a PR for it there. 
I am not going to merge a PR with a full exchange API implementation to Freqtrade, since I do not intend to support that going forward - since I know it's a lot of work.

**Comment by creslinux** [comment]
"Sub classing will make life easier".
This conversation has surface many times in the last 6 months that I'm
aware of.

I don't see FT implementing this as doing CCXTs work.
CCXT provides functionality and documentation to explicitly pass exchange
specific argument. It's expected to be used to make use of CCXT
effectively.

Subclassing as I see for FT is simply a means to abstract one, as example,
_set_stop_loss() exchange method to populate its signature from another.
It is an impossibly difficult task and therefore unlikely useful
expectation that CCXT can unify all call types across all exchanges.

Subclassing/abstracting none unified CCXT calls for popular exchanges is
eminently sensible for FT in my view.
The introduction of a factory for users to pick an actual subclass to use
would also be helpful to be more modular, be easier to support user
contribution.

My two cents worth.

**Comment by mishaker** [comment]
@xmatthias 
>If a exchange is missing from ccxt, it's honestly better to submit a PR for it there.
I am not going to merge a PR with a full exchange API implementation to Freqtrade, since I do not intend to support that going forward - since I know it's a lot of work.

We are on the same page :) 

@creslinux 
Yep you are right, and subclassing will be merged soon.

**Comment by xmatthias** [comment]
the first subclass is merged since this morning ... :)

**Comment by hroff-1902** [comment]
> I've heard that's nearly impossible...

I don't know why do you have such opinion, I interact with him making changes in ccxt that would help this bot working with BitMEX without any subclassing and rusty nails, and can say he's quite responsive... 

)))) :laughing:

**Comment by xmatthias** [comment]
you know that that comment was made by kroitor himself - yes?

**Comment by hroff-1902** [comment]
> you know that that comment was made by kroitor himself - yes? smiley

Hmm, don't you see the laughing emoji there? :thinking:

**Comment by iuvbio** [comment]
@kroitor 
 > I've heard that's nearly impossible...

I see it's very difficult indeed :laughing:  

> Not an easy thing to do, though...

Absolutely, I appreciate that and all the work you're doing :+1: 

@creslinux 
> Subclassing as I see for FT is simply a means to abstract one, as example, _set_stop_loss() exchange method to populate its signature from another. It is an impossibly difficult task and therefore unlikely useful expectation that CCXT can unify all call types across all exchanges.

I agree to a certain extend. Obviously, I'm definitely not against sublassing. I just think, for example for sublass orders, leverage, etc., there is definitely the possibility for it to be unified in ccxt, as kroitor says himself. That's not to say that we shouldn't implement features that are not unified there yet. Rather I mean that **ideally** ccxt could take over the logic for some of the unification that we define. We could take some of the work off ccxt's shoulders so to say. I know that ccxt is written mainly in java and then transpiled to python and php, but porting the logic should be quite straightforward. All that is of course assuming that we even implement a significant number of exchanges, as (imho) our focus should not primarily be on fully supporting the maximum number of exchanges, but to provide robust and advanced trading capabilities. Of course, on the more exchanges the better.

> The introduction of a factory for users to pick an actual subclass to use would also be helpful to be more modular, be easier to support user contribution.

Fully agree on that.

@xmatthias 
> yes - i was thinking of implementing something similar to ccxt's `has` dict - (maybe `ft_has`) - which defines which subclass supports what feature (can be right next to the checks we do against the `has` dict...

Sounds good :+1: 

Also agree on the API, we shouldn't try to implement exchanges that are not supported by ccxt at all.

@everyone
I really like #1582, I think going forward there a few more things to be done. Mostly probably decide on what the "generic" class should be. Only things that are supported by all exchanges? Or do we take one exchange as the default, as currently bittrex. I can see a problem there, when we start to add other bittrex specific stuff and then the generic class is instantiated with an unsupported one that doesn't have that functionality. We should probably define all methods in the base class and raise a not supported exception when not implemented, as it is in ccxt.

I think this issue is getting a bit overloaded, should we move multiple exchange support to a different issue? Or an issue for each exchange and make a multiple exchange project or something like that?

**Comment by xmatthias** [comment]
ideally, the base Exchange class only has things ccxt unifies.
Using `ccxt.has` checking for what is actually supported is easy.
the above is not the case yet (binance subclass is merged into the base-class...).

I would not focus on implementing, or even supporting as many exchanges as possible, but focus on a few key ones, and working reliably on these.
IMHO, there is no benefit in supporting 10 exchanges. 
It'll slow overall movement of the project down since a lot of focus and time needs to go to debug/fix exchange-specific bugs. Ccxt takes most of it from us - but not if we add too much exchange-specific code. 

The focus should be on managing the lifecycle of the bot - having reliable fallback methods (ideally checks beforehand) if creating certain orders (sell?) fail.

Maybe having a good documentation (step by step guide) on how to write and test functions.

**Comment by kroitor** [comment]
Hello guys! Your suggestions, proposals, opinions and comments are very welcome here: https://github.com/ccxt/ccxt/issues/4834 )

**Comment by xmatthias** [comment]
I'm going to close this issue now because all points discussed above have been implemented as far as i can see:

* Exchange subclasses
* dataprovider
* dynamic whitelists
* Order handler (FreqtradeBot.py  - but it really doesn't do much else now since the main handling moved to `worker.py`)

**Comment by hroff-1902** [comment]
marked it as it was an 'ongoing discussion' in its time, for find it easier
it's not ongoing anymore, so closed, it's okay I guess
